enhancements/sha-bundle-digest.md

```diff
@@ -0,0 +1,136 @@
+feature and its implications. This can be mitigated with clear documentation both on the website and in the CLI help
+section.
+
+## Design Details
+
+- As of this EP, the new `--use-image-digests` is the only flag to pass to the `operator-sdk generate bundle` command.
+  In order to allow users full control over the command's flags, the `Makefile` should use an intermediary variable to
+  hold the flags that a user can overwrite entirely. For example:
+
+```Makefile
+GENERATE_BUNDLE_FLAGS ?= ""
+USE_IMAGE_DIGESTS ?= false
+ifeq ($(GENERATE_BUNDLE_FLAGS),"")
+    ifeq($(USE_IMAGE_DIGESTS),true)
+        GENERATE_BUNDLE_FLAGS += "--use-image-digests"
+    endif
+endif
+bundle:
+    operator-sdk generate bundle $(GENERATE_BUNDLE_FLAGS)
+```
+
+- The `RELATED_IMAGE_*` variables should be defined in the `manager.yaml` file as outlined
+  [here](https://master.sdk.operatorframework.io/docs/best-practices/common-recommendation/#other-common-suggestions).
+- In order to resolve digests from tags, the CLI will use the `imageresolver` package in
+  [operator-manifest-tools](https://github.com/operator-framework/operator-manifest-tools).
+- The output cluster CSV will have the `relatedImage` annotation that is explained in the
+  [OpenShift Documentation](https://docs.openshift.com/container-platform/4.9/operators/operator_sdk/osdk-generating-csvs.html#olm-enabling-operator-for-restricted-network_osdk-generating-csvs).
+
+### Test Plan
+
+The resulting artifact of this EP will be a cluster CSV with image digests instead of tags in the image definitions as
+well as digest annotations. This feature will need an integration test that veifiesthat the knobs we provide to enable
+and configure image digests generates an expected CSV. The expected CSV will be based off of the existing memcached CSV,
+located
+[here](https://github.com/operator-framework/operator-sdk/blob/master/testdata/go/v3/memcached-operator/bundle/manifests/memcached-operator.clusterserviceversion.yaml)
+with the `spec.relatedImages` section from the proposal section added to it.
+
+## Implementation History
+
+There has been no implementation to this point.
+
+## Drawbacks
+
+The original motivation for this feature is used in a disconnected environment with an SDK-built operator, OLM, and OCP.
+Performing a full E2E test in a complex environment that will need to be specifically built out for this task is out of
+scope for this SDK EP.
+
+## Alternatives
+
+We are not considering any alternative solutions at this time.
+
+## Open Questions
+
+- Operator-manifest-tools relies on an external binary, skopio. In order to remove the external binary, can we PR
+  operator-manifest-tools to replace the calls to the skopeo CLI with library calls?
+- Store digest and tag mappings in file so that user-written code can resolve tags to digest.
+- What is the function of `spec.relatedImage.name`?
+- How are the values of `spec.relatedImage.name` used?
```

Comment:
```suggestion
- How are the values of `spec.relatedImage.name` used?

## Appendix A

The following is an example of an image manifest for `golang:1.17.6`:

```json
[
  {
    "Config": "8b86bf336a01235faf28137dae90772076e6f431a2951259d949eb9153012755.json",
    "RepoTags": [
      "golang:1.17.6"
    ],
    "Layers": [
      "644b6e1dc104cf7c01e3d906bad42ca6386e62e03ba55c23c9229eeca2aedf4e/layer.tar",
      "dbaefc91c060fe8e827580d1018824c5479bb8eeadf644b44f9bbdc6cc0ad314/layer.tar",
      "7164fcc370aec09c29fb107f4b404d667a046d7eb652130cfcad3b493d7f9485/layer.tar",
      "ded273d40fe1a66f4cc0e6cc15236a8800c4e465df4df2b682422a5477910951/layer.tar",
      "eac35aa9198a17ca6942cda2efacf8b052fcfd7c9046cace7290a0b80bd43d7b/layer.tar",
      "34cd146535bd6d8585e4bc411593bc82a336a335fbcc0a1cf78b1ecfe635c549/layer.tar",
      "0aafe4d111d6dbd2414e183de18f5736f470de5b0e750b05bbe4063f00366386/layer.tar"
    ]
  }
]

The manifest is built out of a list of layers, repo tags, and a config file. Since the config file and layers themselves
are hashed, if any of them change or the repo tags, the manifest digest will change.

## Appendix B

A `ClusterServiceVersion` is a [CRD](https://olm.operatorframework.io/docs/concepts/crds/clusterserviceversion/) defined by [Operator Lifecycle Manager](https://github.com/operator-framework/operator-lifecycle-manager) that defines metadata about the entire operator.
```

An example CSV for the Memcached example operator is available [here](https://github.com/operator-framework/operator-sdk/blob/master/testdata/go/v3/memcached-operator/bundle/manifests/memcached-operator.clusterserviceversion.yaml).

This EP will change two things on the CSV: use digests in deployment container image locations and add the `spec.relatedImages` section with all other images, such as base images, that need to have a digest resolved from a tag. An abbreviated version of the CSV with the aforementioned changes looks like this:

```yaml
apiVersion: operators.coreos.com/v1alpha1
kind: ClusterServiceVersion
metadata:
  # ...
spec:
  # ...
  install:
    spec:
      # ...
      deployments:
      - name: memcached-operator-controller-manager
        spec:
          # ...
          template:
            # ...
            spec:
              containers:
              - args:
                  # ... 
                image: gcr.io/kubebuilder/kube-rbac-proxy@sha256:e10d1d982dd653db74ca87a1d1ad017bc5ef1aeb651bdea089debf16485b080b
                # ...
              - args:
                  # ... 
                image: quay.io/example/memcached-operator@sha256:d044b878fbcbd0c06c2bcc050f3f9ca070ffb089e6bac01accddcd0ba85b7b9e
                # ...
      # ...
    # ...
  # ...
  relatedImages:
  - name: golang:1.17
    image: # TODO: fill out this
  - name: gcr.io/distroless/static:nonroot
    image: # TODO: fill out this
```

## Appendix C

TODO(ryantking): Example and explanation of manager.yaml file. 
```

Reply:
Added manager yaml sample :+1: